lib/pure/streams.nim

```diff
@@ -241,19 +244,29 @@ proc readDataStr*(s: Stream, buffer: var string, slice: Slice[int]): int =
     # fallback
     result = s.readData(addr buffer[0], buffer.len)
 
-when not defined(js):
+template jsOrVm(caseJsOrVm, caseElse: untyped): untyped =
+  when nimvm:
+    block:
```

Comment:
why block?

Reply:
Because you can't do this:

```nim
when nimvm:
  var foo = 3
else:
  var foo = 4 # redefinition
```

If you remove the `block` and do the following:

```nim
jsOrVm:
  var foo = 3
do:
  var bar = 4
```

This expands to:

```nim
when nimvm:
  var foo = 3
else:
  when defined(js):
    var foo = 3
  else:
    var bar = 4
```

So if JS is defined, you get a redefinition error.

Reply:
my point remains; you're changing semantics, so you're increasing risk of code that works with js+c and works differently with nimvm (say, after someone else re-uses your template)
`when nimvm` can't be used for that: https://nim-lang.github.io/Nim/manual.html#statements-and-expressions-when-nimvm-statement

> Code in branches must not affect semantics of the code that follows the when nimvm statement. E.g. it must not define symbols that are used in the following code.

Reply:
What should I do here then? Put the `else` branch in a `block` too and change the name to `jsOrVmLocal`?

Reply:
```nim
template jsOrVm(caseJsOrVm, caseElse: untyped): untyped =
  when nimvm:
    block: caseJsOrVm
  else:
    block:
      when defined(js): caseJsOrVm
      else: caseElse
```
it's hard for me to test because `nim js -r tests/js/tstreams.nim` fails for me regardless of this change [1] but at least it compiles with this change.

[1]
`nim js -r tests/js/tstreams.nim`
```
/Users/timothee/git_clone/nim/Nim_temp3/tests/js/tstreams.js:692
    throw new Error(cbuf_3715201);
    ^

Error: Error: unhandled exception: index 0 not in 0 .. NaN [IndexError]
Traceback (most recent call last)
streams.peekChar
streams.peekData, line: 303
streams.ssPeekData, line: 1156
[]=.[]=, line: 2463
```

but hopefully https://github.com/timotheecour/Nim/issues/124 can be fixed and we'll be able to introduce your current version of `jsOrVm`

Reply:
I made this change already but named  `jsOrVmLocal`, will rename to `jsOrVmBlock` next commit, marking as resolved